### ISSUE TITLE
Problem with Batched Inference in example.py script

### ISSUE THREAD

**Issue by michaelholm-ce** [NVIDIA/DIGITS#536]
I may be missing something here, or I have found a bug.

I believed that changing the batch_size to something greater than 1 (i.e. batched inference) should give the same individual scores, just maybe complete the task faster.  However, I have found that changing the batch_size actually changes the resulting scores (they are no longer correct).

Am I in need of a deeper lesson on neural networks, or have I found a bug?

Thanks,

Michael

**Comment by lukeyeager** [comment]
I assume you're talking about the script at `examples/classification/example.py`.

Are you running into this bug - #409? Which version of DIGITS do you have?
```sh
$ ./digits-devserver --version
3.1-dev
```

**Comment by michaelholm-ce** [comment]
I'm running the script at examples/classification/example.py with digits version 3.1-dev.

I did read issue #409 and made the deep copy scores change, but the problem persists.

**Comment by michaelholm-ce** [comment]
More specifically, I changed the example.py script only in the following way:

I replaced "image_files = [args['image']]" with 
 im_path = args['image']
 if os.path.isdir(im_path):
      image_files = [os.path.join(im_path,f) for f in os.listdir(im_path)]
  elif os.path.isfile(im_path):
      image_files = [im_path]

I have a directory with N images.  When I run the images individually, I get the correct scores.  When I run the whole directory, the first image scores correctly and the remainder score incorrectly.

**Comment by gheinrich** [comment]
I am wondering why you had to change the code to to a deep copy of scores since the change should already be there on 3.1-dev? (see this [line](https://github.com/NVIDIA/DIGITS/blob/master/examples/classification/example.py#L143)). Or am I missing anything?

**Comment by michaelholm-ce** [comment]
You're right @gheinrich -- my mistake.  The truth is that I first saw the problem when working with an older version (when I did make the deep copy change).  When the problem persisted, I updated to version 3.1-dev (where I made the other changes listed above) and the problem persisted.  Thanks for pointing that out.

**Comment by gheinrich** [comment]
I tried making the same changes as you did however it feels to me like predictions are correct. Did you try using `use_archive.py`?

**Comment by michaelholm-ce** [comment]
To clarify @gheinrich , are you saying that you ran `example.py` on a full directory (even as small as two images), and the output scores were identical to the scores you get when you run `example.py` on the images individually?

I have used `use_archive.py` in the past.  However, there was a minor change in the `deploy.prototxt` format among versions, and since `use_archive.py` takes a `.tar` file, that always causes an error.  They both call the same `classify` function, though, so I don't see that making any difference.

**Comment by gheinrich** [comment]
Yes, I pushed a version of the code that includes your change in `use_archive.py` [there](https://github.com/gheinrich/DIGITS/tree/dev/use-archive-with-dir).

I have created a directory with images from the MNIST dataset. If I run the script on this directory then all images are correctly classified with high confidence:

```
$ ./use_archive.py ~/Downloads/20160117-135614-955c_epoch_5.0.tar.gz images/
Unknown file: train_val.prototxt
Unknown file: solver.prototxt
Processed 5/5 images ...
Classification took 0.00620698928833 seconds.
----------------------- Prediction for images/three.png ------------------------
 99.9999% - "3"
  0.0001% - "8"
  0.0000% - "7"
  0.0000% - "9"
  0.0000% - "2"
------------------------ Prediction for images/five.png ------------------------
 81.0012% - "5"
 10.7394% - "6"
  8.1783% - "8"
  0.0762% - "9"
  0.0022% - "3"
------------------------ Prediction for images/four.png ------------------------
 99.8313% - "4"
  0.1677% - "9"
  0.0007% - "7"
  0.0001% - "8"
  0.0000% - "1"
------------------------ Prediction for images/one.png -------------------------
 99.7627% - "1"
  0.1565% - "7"
  0.0508% - "4"
  0.0170% - "8"
  0.0072% - "2"
------------------------ Prediction for images/zero.png ------------------------
 99.8964% - "0"
  0.0868% - "6"
  0.0103% - "2"
  0.0051% - "9"
  0.0004% - "8"
Script took 0.253932952881 seconds.
```

**Comment by michaelholm-ce** [comment]
Hello @gheinrich.  Thank you for running the test.  I see that in your test, the five image files were classified correctly.  Will you please take the test one step further and verify that you get the same scores (or very close) when you run the image files individually?

I am still quite concerned that there is something wrong here.  I checked out the branch you supplied, used a caffe model I trained with digits showing ~98% accuracy on the validation set, and tested on color images that are the exact size expected by the model (256x256).  In every case, the first image in a batch is classified correctly (and with identical score as when scored individually), but all the other test images in the directory have different scores.  When I change `batch_size` to 1 (as opposed to the default 25) in the `forward_pass` function, then all images are scored correctly, but this substantially slows down classification speed.

**Comment by gheinrich** [comment]
On 4825f8de2374904139ca5ea3fd5a7df310f0b6d6 I get exactly the same predictions when I classify images one by one:
```
(venv)greg@greg:~/ws/digits/examples/classification$ ./use_archive.py ~/Downloads/20160117-135614-955c_epoch_5.0.tar.gz images/three.png 
Unknown file: train_val.prototxt
Unknown file: solver.prototxt
Processed 1/1 images ...
Classification took 0.00297689437866 seconds.
----------------------- Prediction for images/three.png ------------------------
 99.9999% - "3"
  0.0001% - "8"
  0.0000% - "7"
  0.0000% - "9"
  0.0000% - "2"

Script took 0.312922000885 seconds.
(venv)greg@greg:~/ws/digits/examples/classification$ ./use_archive.py ~/Downloads/20160117-135614-955c_epoch_5.0.tar.gz images/five.png 
Unknown file: train_val.prototxt
Unknown file: solver.prototxt
Processed 1/1 images ...
Classification took 0.00299596786499 seconds.
------------------------ Prediction for images/five.png ------------------------
 81.0012% - "5"
 10.7394% - "6"
  8.1782% - "8"
  0.0762% - "9"
  0.0022% - "3"

Script took 0.267504930496 seconds.
(venv)greg@greg:~/ws/digits/examples/classification$ ./use_archive.py ~/Downloads/20160117-135614-955c_epoch_5.0.tar.gz images/four.png 
Unknown file: train_val.prototxt
Unknown file: solver.prototxt
Processed 1/1 images ...
Classification took 0.00307488441467 seconds.
------------------------ Prediction for images/four.png ------------------------
 99.8313% - "4"
  0.1677% - "9"
  0.0007% - "7"
  0.0001% - "8"
  0.0000% - "1"

Script took 0.2662088871 seconds.
(venv)greg@greg:~/ws/digits/examples/classification$ ./use_archive.py ~/Downloads/20160117-135614-955c_epoch_5.0.tar.gz images/one.png 
Unknown file: train_val.prototxt
Unknown file: solver.prototxt
Processed 1/1 images ...
Classification took 0.00297904014587 seconds.
------------------------ Prediction for images/one.png -------------------------
 99.7627% - "1"
  0.1565% - "7"
  0.0508% - "4"
  0.0170% - "8"
  0.0072% - "2"

Script took 0.259674787521 seconds.
(venv)greg@greg:~/ws/digits/examples/classification$ ./use_archive.py ~/Downloads/20160117-135614-955c_epoch_5.0.tar.gz images/zero.png 
Unknown file: train_val.prototxt
Unknown file: solver.prototxt
Processed 1/1 images ...
Classification took 0.00302720069885 seconds.
------------------------ Prediction for images/zero.png ------------------------
 99.8964% - "0"
  0.0868% - "6"
  0.0103% - "2"
  0.0051% - "9"
  0.0004% - "8"

Script took 0.258541107178 seconds.
```

**Comment by michaelholm-ce** [comment]
Thank you, @gheinrich, for running the test. 

I'm at a loss as to where to go from here.  I checked out the digits commit you shared above (4825f8d) and still get garbage output (for all images after the first) when I run against a directory when `batch_size` is greater than 1.

**Comment by gheinrich** [comment]
I used the canonical LeNet, trained on MNIST. The 5 images are used were randomly picked from the `test` folder. Or would you like me to upload the binary files (where)?

**Comment by lukeyeager** [comment]
@michaelholm-ce thanks for the bug report, this does seem to be a legit bug!

1. You are using cuDNN v4 on a [Maxwell GPU](https://en.wikipedia.org/wiki/CUDA#Supported_GPUs), is that correct?
2. If you increase the number of images in your batch (e.g. from 5 to 30), do you start getting the right answers again?

**Comment by lukeyeager** [comment]
If anyone else runs into this issue, try upgrading from the cuDNN 4 release candidate (4.0.4) to the final release (4.0.7).
